pmd-core/src/main/java/net/sourceforge/pmd/ruledef/RuleFactory.java

```diff
@@ -0,0 +1,412 @@
+                    setPropertyValues(ruleReference, (Element) node);
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unexpected element <" + node.getNodeName()
+                                                           + "> encountered as child of <rule> element for Rule "
+                                                           + ruleReference.getName());
+
+                }
+            }
+        }
+
+
+        return ruleReference;
+    }
+
+
+    /**
+     * Parses a rule element and returns a new rule instance.
+     *
+     * <p>Notes: The ruleset name is not set here. Exceptions raised from this method indicate invalid XML structure,
+     * with regards to the expected schema, while RuleBuilder validates the semantics.
+     *
+     * @param ruleElement The rule element to parse
+     *
+     * @return A new instance of the rule described by this element
+     *
+     * @throws IllegalArgumentException if the element doesn't describe a valid rule.
+     */
+    public Rule buildRule(Element ruleElement) {
+
+        checkRequiredAttributesArePresent(ruleElement);
+
+        String name = ruleElement.getAttribute(NAME.name);
+
+        RuleBuilder builder = new RuleBuilder(name,
+                                              ruleElement.getAttribute(CLASS.name),
+                                              ruleElement.getAttribute(LANGUAGE.name));
+
+
+        if (ruleElement.hasAttribute(MINIMUM_LANGUAGE_VERSION.name)) {
+            builder.minimumLanguageVersion(ruleElement.getAttribute(MINIMUM_LANGUAGE_VERSION.name));
+        }
+
+        if (ruleElement.hasAttribute(MAXIMUM_LANGUAGE_VERSION.name)) {
+            builder.maximumLanguageVersion(ruleElement.getAttribute(MAXIMUM_LANGUAGE_VERSION.name));
+        }
+
+        if (ruleElement.hasAttribute(SINCE.name)) {
+            builder.since(ruleElement.getAttribute(SINCE.name));
+        }
+
+        builder.message(ruleElement.getAttribute(MESSAGE.name));
+        builder.externalInfoUrl(ruleElement.getAttribute(EXTERNAL_INFO_URL.name));
+        builder.setDeprecated(hasAttributeSetTrue(ruleElement, DEPRECATED.name));
+        builder.usesDFA(hasAttributeSetTrue(ruleElement, DFA.name));
+        builder.usesTyperesolution(hasAttributeSetTrue(ruleElement, TYPERESOLUTION.name));
+        builder.usesMetrics(hasAttributeSetTrue(ruleElement, METRICS.name));
+
+
+        Element propertiesElement = null;
+
+        final NodeList nodeList = ruleElement.getChildNodes();
+        for (int i = 0; i < nodeList.getLength(); i++) {
+            Node node = nodeList.item(i);
+            if (node.getNodeType() != Node.ELEMENT_NODE) {
+                continue;
+            }
+            RulesetSchemaConstants nodeName = EnumUtils.getEnum(RulesetSchemaConstants.class,
+                                                                node.getNodeName().toUpperCase());
+
+            switch (nodeName) {
```

Comment:
Yeah actually switch on string is available from Java 7. It may seem dumb but I made an enum because Checkstyle complained about string literals being repeated x) And I find enums cleaner than a bunch of `private static final String` constants. Though to be honest we could inline everything and use the literals, they're repeated two or three times max